### PR TITLE
Removedoubles

### LINKED ISSUE #271: Replace all doubles with Nums
in all indicator and criteria calculation.  Agree or disagree?

### PR DESCRIPTION
**Pull request by edlins**
Fixes #271 #217 #276 .

Changes proposed in this pull request:
- same as #276 but with new RuntimeException in PrecisionNum.sqrt

- [x] added an entry to the unreleased section of `CHANGES.md`

### CODE REVIEW

**Comment by edlins** [comment]
I fully expect there to be discussion regarding the changes to LTCC.  It feels messy and probably begs for some redesign.

**Comment by team172011** [comment]
Does this replace #276  ?

File: ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/LinearTransactionCostCriterion.java [hunk 1]
```diff
@@ -124,7 +127,7 @@ private Num getOrderCost(Order order, Num tradedAmount) {
      * @return the absolute total cost of all orders in the trade
      */
     private Num getTradeCost(TimeSeries series, Trade trade, Num initialAmount) {
-        Num totalTradeCost = series.numOf(0);
+        Num totalTradeCost = SeriesBuilder.getDefaultFunction().apply(0);
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/LinearTransactionCostCriterion.java]: So LTCC calculates only with the defaultFunction of `SeriesBuilder`? What if the `numFunction ` of `TimeSeries ` parameter is different? My first thought is to drop the Constructor initialization and only provide static `LTCC #calculate `functions with additional parameter `a`, `b` and `initialAmount` (here we have a time series with corresponding `numFunction`)

File: ta4j-core/src/main/java/org/ta4j/core/num/PrecisionNum.java [hunk 2]
```diff
@@ -297,8 +297,7 @@ public Num sqrt(int precision) {
             try {
                 estimate = (BigDecimal) format.parse(estimateString);
             } catch (ParseException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
+                throw new RuntimeException("Unable to parse into DecimalFormat the following string:" + estimateString + ": " + e);
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/num/PrecisionNum.java:300]: looks good

**Comment by edlins** [comment]
> Does this replace #276 ?

#276 was already merged.  Wasn't sure if you wanted something new or not.

File: ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/LinearTransactionCostCriterion.java (hunk 1 above)
Comment (edlins) [reply ta4j-core/src/main/java/org/ta4j/core/analysis/criteria/LinearTransactionCostCriterion.java]: I don't remember now why I made this change.  I think it had to do with a `ClassCastException` but when I undo that one line the tests run fine.  So I've changed it back.

File: ta4j-core/src/test/java/org/ta4j/core/analysis/criteria/LinearTransactionCostCriterionTest.java [hunk 3]
```diff
@@ -38,7 +38,7 @@
     private ExternalCriterionTest xls;
 
     public LinearTransactionCostCriterionTest(Function<Number, Num> numFunction) throws Exception {
-        super((params) -> new LinearTransactionCostCriterion((double) params[0], (double) params[1], (double) params[2]),numFunction);
+        super((params) -> new LinearTransactionCostCriterion((double) params[0], (double) params[1], (double) params[2], numFunction), numFunction);
```
Comment (edlins) [inline ta4j-core/src/test/java/org/ta4j/core/analysis/criteria/LinearTransactionCostCriterionTest.java:41]: This is starting to feel hideous.  May need a redesign on how `numFunctions` are handled.  They are now
1) stored as an API default
2) stored in TimeSeries
3) passed explicitly

Maybe 2) and 3) should be removed and only 1) used..?  Regardless it feels like it's getting unwieldy.

**Pull request closed without merging**